Incremental recovery and rebalance should run pg_rewind in parallel

asimrp opened the pull request:
Incremental recovery and rebalance operations involve running pg_rewind against failed primaries.  This patch changes gprecoverseg such that pg_rewind is invoked in parallel, using the WorkerPool interface, for each affected segment in the cluster.  There is no reason to rewind segments one after the other.

Fixes Github issue #9466

Let's see how PR pipeline likes it.

paul-guo- commented on gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:
```diff
@@ -362,19 +364,32 @@ def run_pg_rewind(self, rewindInfo):
+            if not cmd.was_successful():
+                dbid = cmd.name.split(':')[1].strip()
+                self.__logger.debug("%s failed" % cmd.name)
+                self.__logger.warning("Incremental recovery failed for dbid %s. You must use gprecoverseg -F to recover the segment." % cmd.name)
```
s/cmd.name/dbid/?

paul-guo- commented on gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:
```diff
@@ -362,19 +364,32 @@ def run_pg_rewind(self, rewindInfo):
+            base.join_and_indicate_progress(self.__pool)
+
+        for cmd in self.__pool.getCompletedItems():
+            self.__logger.debug('pg_rewind results: %s' % cmd.results)
```
Not sure what is output of cmd.results. Probably this should include dbid or target info.

marksliva replied:
It will render using `CommandResult.__str__()` which includes the return code, stdout and stderr.
To include the `dbid`, you can render the entire command, which will include the command name (which contains the dbid), command that was run, and results.
`self.__logger.warning('pg_rewind command: %s' % cmd)`
```
class Command
...
    def __str__(self):
        if self.results:
            return "%s cmdStr='%s'  had result: %s" % (self.name, self.cmdStr, self.results)
        else:
            return "%s cmdStr='%s'" % (self.name, self.cmdStr)
```

asimrp commented on gpMgmt/bin/gppylib/commands/base.py:
```diff
@@ -288,7 +288,7 @@ def run(self):
 
     def haltWork(self):
-        self.logger.debug("[%s] haltWork" % self.name)
+        self.logger.debug("[%s] haltWork: %s" % (self.name, self.cmd))
```
This change was for debugging, I will omit it.

asimrp commented:
Snippet of debug output changed by this patch:
```
20200206:12:27:27:010679 gprecoverseg:asimmac:apraveen-[DEBUG]:-[worker0] finished cmd: rewind dbid: 2 cmdStr='ssh -o StrictHostKeyChecking=no -o ServerAliveInterval=60 asimmac ". /Users/apraveen/workspace/gpdb/gpAux/greenplum-db-devel/greenplum_path.sh; [ -f /Users/apraveen/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast1/demoDataDir0/recovery.conf ] || PGOPTIONS=\"-c gp_session_role=utility\" $GPHOME/bin/pg_rewind --write-recovery-conf --slot=\"internal_wal_replication_slot\" --source-server=\"host=asimmac port=7005 dbname=template1 application_name=__gprecoverseg_pg_rewind__\" --target-pgdata=/Users/apraveen/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast1/demoDataDir0 --progress 2>&1"'  had result: cmd had rc=0 completed=True halted=False
  stdout='...'
20200206:12:27:30:010679 gprecoverseg:asimmac:apraveen-[DEBUG]:-pg_rewind results: cmd had rc=0 completed=True halted=False
  stdout='...'
```

paul-guo- commented:
In the github issue, it's 300+ node recovery. Did not check worker pool code, but I'm curious if worker pool works fine for that many parallel workers?

marksliva commented on gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:
```diff
@@ -362,19 +364,32 @@ def run_pg_rewind(self, rewindInfo):
 
             # Run pg_rewind to do incremental recovery.
-            cmd = gp.SegmentRewind('segment rewind',
+            assert dbid == targetSegment.getSegmentDbId()
```
Why are we checking this? We do not reload the target segment between the time the value is read and compared, so these 2 values will always be equal, since `dbid` comes from `target.getSegmentDbId()`.

asimrp replied:
That serves to clarify what the newly added argument to `run_pg_rewind()` should adhere to.  We can remove it if it seems out of place, I don't feel strongly about it.

marksliva replied:
I don't see this becoming an issue. For it to get out of sync, someone would have to mutate `targetSegment` or the map key/value pair.

marksliva commented on gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:
```diff
@@ -270,7 +270,8 @@ def buildMirrors(self, actionName, gpEnv, gpArray):
         mirrorsToStart = []
-        rewindInfo = []
+        # Map of dbid to tuples of the form (GpMirrorToBuild, src host, src port)
+        rewindInfo = {}
```
This map is fine, although I am curious why it was changed? We could call `getSegmentDbId()` on the target segment object that gets passed in instead of changing the data structure.

marksliva replied:
Oh now I see how it gets used below when rendering errors. No need to change.

marksliva commented on gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:
```diff
@@ -362,19 +364,32 @@ def run_pg_rewind(self, rewindInfo):
+        for cmd in self.__pool.getCompletedItems():
+            self.__logger.debug('pg_rewind results: %s' % cmd.results)
+            if not cmd.was_successful():
+                dbid = cmd.name.split(':')[1].strip()
```
This will evaluate to a string i.e. `'2'`. The `rewindInfo` keys are integers, which will throw a `KeyError` unless you parse it as an int: `int(cmd.name.split(':')[1].strip())`.

asimrp replied:
Thank you for the catch!  I've made the change and am working on a unit test to exercise that code path.

asimrp commented:
@paul-guo- the WorkerPool object used by this patch is created by the user of `GpMirrorListToBuild` class, which is `GpRecoverSegmentProgram`.  Thus the batch size for it is set according to `-B` command line option in `GpRecoverSegmentProgram.run()`.

asimrp commented:
Thank you for the reviews so far.  I've incorporated the comments and also added a unittest in a separate commit.  The unittest never finishes, I'm not able to figure out why.  Can you please help?  I'm running it as:
```
asimmac:~/workspace/gpdb/gpMgmt/bin$ python -m unittest gppylib.test.unit.test_unit_buildmirrorsegments
20200207:19:54:52:006340 python -m unittest:default-[WARNING]:-Incremental recovery failed for dbid 2. You must use gprecoverseg -F to recover the segment.
20200207:19:54:52:006340 python -m unittest:default-[WARNING]:-Incremental recovery failed for dbid 3. You must use gprecoverseg -F to recover the segment.
20200207:19:54:52:006340 python -m unittest:default-[WARNING]:-Incremental recovery failed for dbid 4. You must use gprecoverseg -F to recover the segment.
20200207:19:54:52:006340 python -m unittest:default-[WARNING]:-Incremental recovery failed for dbid 5. You must use gprecoverseg -F to recover the segment.
.
----------------------------------------------------------------------
Ran 1 test in 0.037s

OK

--> it's getting stuck at this point <--
```

marksliva commented:
> Thank you for the reviews so far. I've incorporated the comments and also added a unittest in a separate commit. The unittest never finishes, I'm not able to figure out why. Can you please help? I'm running it as:
> 
> ```
> asimmac:~/workspace/gpdb/gpMgmt/bin$ python -m unittest gppylib.test.unit.test_unit_buildmirrorsegments
> 20200207:19:54:52:006340 python -m unittest:default-[WARNING]:-Incremental recovery failed for dbid 2. You must use gprecoverseg -F to recover the segment.
> 20200207:19:54:52:006340 python -m unittest:default-[WARNING]:-Incremental recovery failed for dbid 3. You must use gprecoverseg -F to recover the segment.
> 20200207:19:54:52:006340 python -m unittest:default-[WARNING]:-Incremental recovery failed for dbid 4. You must use gprecoverseg -F to recover the segment.
> 20200207:19:54:52:006340 python -m unittest:default-[WARNING]:-Incremental recovery failed for dbid 5. You must use gprecoverseg -F to recover the segment.
> .
> ----------------------------------------------------------------------
> Ran 1 test in 0.037s
> 
> OK
> 
> --> it's getting stuck at this point <--
> ```
@asimrp 
The problem is that threads which were started by the `WorkerPool` are still running. Stopping the workers at the end of each test fixes the problem you're seeing.
```
    def setUp(self):
        self.worker_pool = WorkerPool()

    def tearDown(self):
        super(GpMirrorListToBuildTestCase, self).tearDown()
        self.worker_pool.haltWork()
        self.worker_pool.joinWorkers()

...
        g = GpMirrorListToBuild(1, self.worker_pool, 1,1)
```

A comment in `test_unit_workerpool.py` explains why this is happening in more depth:
```
        # All background threads must be stopped, or else the test runner will
        # hang waiting. Join the stopped threads to make sure we're completely
        # clean for the next test.
```

marksliva commented on gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:
```diff
@@ -362,19 +364,32 @@ def run_pg_rewind(self, rewindInfo):
+                dbid = int(cmd.name.split(':')[1].strip())
+                self.__logger.debug("%s failed" % cmd.name)
+                self.__logger.warning("Incremental recovery failed for dbid %d. You must use gprecoverseg -F to recover the segment." % dbid)
+                rewindFailedSegments.append(rewindInfo[dbid])
```
This line will cause the program to crash before completing its work. This happens because a tuple is now being appended to `rewindFailedSegments`, and then the line `mirrorsToStart.remove(failedSegment)` fails with `(Reason='list.remove(x): x not in list')`.

This should be changed to:
```
failedSegment = rewindInfo[dbid][0]
rewindFailedSegments.append(failedSegment)
```
It could be useful to create a helper class that wraps these tuple values.
```
class RewindSegmentData:
    def __init__(self, targetSegment, sourceHostname, sourcePort):
        self.targetSegment = targetSegment
        self.sourceHostname = sourceHostname
        self.sourcePort = sourcePort
```
Or consider using a dictionary.
```
{
    "targetSegment": targetSegment
...
}
```

asimrp commented:
@marksliva thank you the patient and thorough review.  Appreciate it very much!  I've fixed the bug and the test.   The test is also changed to make use of the returned list, so as to catch the bug you pointed out.

asimrp commented:
Back ported to 6X_STABLE: 3ddbed8eccb27dddd397848a800778c16cbcdc90
